sys/kern/uio.c

```diff
@@ -86,3 +88,33 @@ int uiomove_frombuf(void *buf, size_t buflen, struct uio *uio) {
 
   return uiomove((char *)buf + offset, buflen - offset, uio);
 }
+
+int uio_init_from_user_iovec(uio_t *uio, uio_op_t op, const struct iovec *u_iov,
+                             int iovcnt) {
+  int error;
+  if (iovcnt <= 0 || iovcnt > IOV_MAX)
+    return EINVAL;
+  const size_t iov_size = sizeof(iovec_t) * iovcnt;
+  iovec_t *k_iov = kmalloc(M_TEMP, iov_size, 0);
+  if ((error = copyin(u_iov, k_iov, iov_size)))
+    goto err_free;
+  size_t len = 0;
+  for (int i = 0; i < iovcnt; i++) {
+    len += k_iov[i].iov_len;
+    /* Ensure that the total data size fits in ssize_t. */
+    if (len > SSIZE_MAX || k_iov[i].iov_len > SSIZE_MAX) {
+      error = EINVAL;
+      goto err_free;
+    }
+  }
+  uio->uio_iov = k_iov;
+  uio->uio_iovcnt = iovcnt;
+  uio->uio_offset = 0;
+  uio->uio_resid = len;
+  uio->uio_op = op;
+  uio->uio_vmspace = vm_map_user();
+  return 0;
+err_free:
+  kfree(M_TEMP, k_iov);
```

Comment:
Because this function is overloaded with responsibility you free `k_iov` here and in `sys_readv` and `sys_writev`.

Reply:
I wouldn't say it's _because_ it's overloaded with responsibility... I'd say stuff like this is necessary whenever a function allocates memory whose ownership is passed to the caller, and the function can fail after allocating the memory, which is the case here.

Reply:
Right. I just find easy to be confused when memory is not allocated and freed at the same lexical level.